BenjaminBeguin/tapas
# Develop

## Test instructions and expectations

> Provide the reviewer with instructions to test your implementation and with what to expect when testing.
First of all you have to run all the basic command line such as : 
```
npm install 
npm run start 
```

you should be able to see the challenges at http://localhost:3000 

## How does it work?

Basically it's a list of user, fetch from an "placeholder" API .
You should be able to drag and drop user to update the list order.
During fetching data, a little loader will be present. Enjoy it. 

I used Axios for fetching data. I know that's an dependency and fetch is included in React. I used that because I used to use it. Very similar to fetch anyway. 
For the drag and drop i used [Dnd library ](http://react-dnd.github.io/react-dnd/docs/api/use-drag) and because I never used it I follow [this tutorial](http://react-dnd.github.io/react-dnd/examples/sortable/simple)
even though the code is very similar i forced myself to understand it. 

I wanted to create animation during drag and drop but i had some issues. 

### Alternatives

I wanted a smooth transition when the list order changed but i had quite some difficulties that i mention in my previous email (was in French, was more easy :D )
I wanted to replace the default ghost preview drag and drop thing with a real DOM element follow the mouse while dragging it. I didn't make it for now but I will feel better when I will find a good way to do it. 
For the transition animation i dont know how to begin with without using absolute position and so long. 

## Critical analysis

To be honest, I not too proud of my code right now. I could be better. Especially for the Drag and drop feature. I followed an example to make it work and while I was reading it, I'm sure that I was not going to think about all the things that guy who make the example think about. 
It was my first time of typescript in a React environment and even thought it's not perfect I'm happy to used it during the test. Same feeling about  JSS. 

### General feeling

Interesting test, challenging test. What I feel after is that I need to learn more about things in development such as Drag and drop behavior and so long. That's what make development interesting. We always have thing to learn. 

> What have you thought about these challenges?

I thing i just replied to this question above 

> Do you think your implementation reflects correctly your technical skills?

As I said, i'm not too proud of my code, and I think that I can do better. 

> (we'll have the opportunity to discuss these points orally, if necessary)

### Difficulties

> If any, list the difficulties you faced when trying to complete the challenges and how you overcome them.

I think again that i already replied to this. But basically i had difficulties with Drag and drop thing. behavior and animation. 
I find solution by looking in the web and by using someone else example. 


### Strengths and weaknesses

> List and argue the points of strengths or weaknesses/oddnesses of your implementation.

**strengths**
 *  Code spliting 
 * I tried to use typescript when it was useful

**weaknesses**
* I don't have the entire control of my code. 

> For weaknesses/oddnesses, suggest improvements and give an estimation of the time necessary to implement them.

* Find a better solution for the drag and drop behavior, build by my own. I think i will need few hour, time to get into the DnD documentation and understand it.
